pg_later/src/api.rs

```diff
@@ -0,0 +1,158 @@
+            Ok(())
+        });
+        ran?
+    }
+    Ok(true)
+}
+
+/// send a query to be executed by the next available worker
+#[pg_extern]
+pub fn pg_later_exec(query: &str) -> Result<i64, spi::Error> {
+    let msg = serde_json::json!({
+        "query": query,
+    });
+    let enqueue = format!("select pgmq_send('pg_later_jobs', '{msg}')");
+    let msg_id: i64 = Spi::get_one(&enqueue)?.expect("failed to send message to queue");
+    Ok(msg_id)
+}
+
+// get the resultset of a previously submitted query
+#[pg_extern]
+fn pg_later_results(job_id: i64) -> Result<Option<pgrx::JsonB>, spi::Error> {
+    let query = format!(
+        "select * from pgmq_pg_later_results
+        where message->>'job_id' = '{job_id}'
+        "
+    );
+    let results: Result<Option<pgrx::JsonB>, spi::Error> = Spi::connect(|mut client| {
+        let mut tup_table: SpiTupleTable = client.update(&query, None, None)?;
+        if let Some(row) = tup_table.next() {
+            let message = row["message"].value::<pgrx::JsonB>()?.expect("no message");
+            return Ok(Some(message));
+        }
+        Ok(None)
+    });
+    let query_resultset = match results {
+        Ok(Some(r)) => r,
+        Ok(None) => {
+            return Ok(None);
+        }
+        _ => {
+            return Err(spi::Error::CursorNotFound(
+                "failed to execute query".to_owned(),
+            ));
+        }
+    };
+    Ok(Some(query_resultset))
+}
+
+// gets a job query from the queue
+pub fn get_job(timeout: i64) -> Option<(i64, String)> {
+    let job = poll_queue(timeout).expect("failed");
+    match job {
+        Some(job) => {
+            let msg_id = job[0].0;
+            let m = serde_json::to_value(&job[0].1).expect("failed parsing jsonb");
+            let q = m["query"].as_str().expect("no query").to_owned();
+            Some((msg_id, q))
+        }
+        None => None,
+    }
+}
+
+fn poll_queue(timeout: i64) -> Result<Option<Vec<(i64, pgrx::JsonB)>>, spi::Error> {
+    let mut results: Vec<(i64, pgrx::JsonB)> = Vec::new();
+
+    let query =
+        format!("select msg_id, message from public.pgmq_read('pg_later_jobs' ,{timeout}, 1)");
+    let _: Result<(), spi::Error> = Spi::connect(|mut client| {
+        let tup_table: SpiTupleTable = client.update(&query, None, None)?;
+        for row in tup_table {
+            let msg_id = row["msg_id"].value::<i64>()?.expect("no msg_id");
+            let message = row["message"].value::<pgrx::JsonB>()?.expect("no message");
+            results.push((msg_id, message));
+        }
+        Ok(())
+    });
+    if results.is_empty() {
+        Ok(None)
+    } else {
+        Ok(Some(results))
+    }
+}
+
+// #[pg_extern]
+// pub fn exec_to_table(
+//     query: &str,
+// ) -> Result<TableIterator<'static, (name!(query, String), name!(results, pgrx::JsonB))>, spi::Error>
+// {
+//     let resultset = query_to_json(query)?;
+//     Ok(TableIterator::new(resultset.into_iter()))
+// }
```

Comment:
```suggestion
```

Reply:
I'm going to keep this one for now since its still in development.